File: jcv-customvalidator-example/src/test/java/com/ekino/oss/jcv/example/jcvcustomvalidatorexample/Case1Test.java
package com.ekino.oss.jcv.example.jcvcustomvalidatorexample;

import java.util.LinkedList;
import java.util.List;

import com.ekino.oss.jcv.core.JsonValidator;
import com.ekino.oss.jcv.core.JsonValueComparator;
import com.ekino.oss.jcv.core.validator.Validators;
import org.junit.jupiter.api.Test;
import org.skyscreamer.jsonassert.ValueMatcherException;

import static com.ekino.oss.jcv.assertion.assertj.JsonCompareAssert.*;
import static com.ekino.oss.jcv.core.validator.Validators.*;
import static com.ekino.oss.jcv.example.jcvcustomvalidatorexample.util.ResourceLoader.*;

public class Case1Test {

    @Test
    void should_validate_json_content_with_custom_validator() {

        assertThatJson(loadJson("case1/actual.json"))
            .using(customValidators())
            .isValidAgainst(loadJson("case1/expected.json"));
    }

    private static List<JsonValidator> customValidators() {
        LinkedList<JsonValidator> validators = new LinkedList<>();

        validators.add(myRefValiadtor());
        validators.addAll(Validators.defaultValidators());

        return validators;
    }

    private static JsonValidator myRefValiadtor() {
        return templatedValidator("my_ref", new MyRefComparator());
    }

    private static class MyRefComparator implements JsonValueComparator<String> {

        @Override
        public boolean hasCorrectValue(String actual, String expected) {
            if (actual != null && actual.startsWith("REF_") && actual.length() == 14) {
                return true;
            }
            throw new ValueMatcherException("Invalid reference format", expected, actual);
        }
    }
}

File: jcv-customvalidator-kotlin-example/src/test/kotlin/com/ekino/oss/jcv/example/jcvcustomvalidatorexample/Case1Test.kt
package com.ekino.oss.jcv.example.jcvcustomvalidatorexample

import com.ekino.oss.jcv.assertion.assertj.JsonCompareAssert.Companion.assertThatJson
import com.ekino.oss.jcv.core.JsonValueComparator
import com.ekino.oss.jcv.core.validator.Validators.defaultValidators
import com.ekino.oss.jcv.core.validator.validators
import org.junit.jupiter.api.Test
import org.skyscreamer.jsonassert.ValueMatcherException

class Case1Test {

    @Test
    fun `should validate json content with custom validator`() {

        assertThatJson(
            // language=json
            """
            {
              "id": "fda7a233-99b9-4756-8ecc-826a1c5a9bf5",
              "reference": "REF_0123456789"
            }
            """.trimIndent()
        )
            .using(validators {
                +defaultValidators()
                +templatedValidator("my_ref", MyRefComparator())
            })
            .isValidAgainst(
                // language=json
                """
                {
                  "id": "{#uuid#}",
                  "reference": "{#my_ref#}"
                }
                """.trimIndent()
            )
    }

    private class MyRefComparator : JsonValueComparator<String> {

        override fun hasCorrectValue(actual: String?, expected: String?): Boolean {
            if (actual != null && actual.startsWith("REF_") && actual.length == 14) {
                return true
            }
            throw ValueMatcherException("Invalid reference format", expected, actual)
        }
    }
}
